### PR TITLE
[WIP] postcss-cssnext merge

### LINKED ISSUE #73: Generic config file
Following #72

What about supporting these file:

- (.)?cssnext.json
- (.)?cssnext.yml
- .cssnextrc (json or yaml)
- (.)?cssnext.config.js (that exports config)

### LINKED ISSUE #93: Compile folder of .css files at once
It will be handy to have such possibility, smth like:

```bash
cssnext --folder ./folder --output-folder ./output
```

### LINKED ISSUE #107: write to a separate source map file from CLI
right now, you can only inline source maps. any way to allow separate source map files?

### LINKED ISSUE #122: Watcher test issues in Linux
So the tests don't run correctly for the watcher code in Linux, I'm either getting:
- The code times out
- The code never finishes

The test in issue is this: test/cli.watcher.js

Killing the processes using the method mentioned here solves the issues:
http://krasimirtsonev.com/blog/article/Nodejs-managing-child-processes-starting-stopping-exec-spawn

However this causes the following line of the test to fail (Which I can't actually see should ever reliably pass):
```
      if (err && !err.killed) { throw err }
```

### LINKED ISSUE #149: Allow to run plugins before transformation (eg: plug a linter, run non-w3c like transformations)
Ref https://github.com/postcss/postcss-import/issues/45
[stylelint](https://github.com/stylelint/stylelint) will be actively supported.

### LINKED ISSUE #156: Reduce overlapping functionality between cssnext & cssnano
It would be good for both of our modules to be able to disable processors based on what has already been transformed. cssnext already disables postcss-calc before the AST goes through to cssnano, but we can do more.

For example, I can detect that automatically by analysing the processor instance for postcss-calc and disabling it silently; same with postcss-pseudoelements.

It would be nice if you could disable postcss-rebeccapurple when cssnano is enabled as it is already converted to hex by postcss-colormin. :smile:

### LINKED ISSUE #173: Define cssnext as a postcss plugin internally
We should follow cssnano api

- cssnext is defined as a postcss plugin (minor change)
- cssnext(options) return a postcss plugin (current behavior)
- cssnext(options).process(string) => result should be the implemented & promoted (new interface)
- maybe we can keep cssnext(string, options) api to avoid a major breaking change (current behavior) - not sure it's a good idea because postcss.plugin might add more change in the future. @ai any comment on that ?

This way, people might use directly cssnext (without braces) as a postcss plugin when https://github.com/postcss/postcss/issues/402 is fixed.

Not a big deal, but I think the reusable instance is a good idea.

Depending if the current api (string, options) can be kept, we might push this as a minor change or as a major bump (2.x)

### LINKED ISSUE #174: pixrem 1.3.0 now accept a browsers option (& output less fallback for IE9/10/11)
It has for now an hardcoded default value to IE8, so we need to pass our value or undefined in order to benefit of default browserslist value in order to disable px fallback when it's not needed.

### LINKED ISSUE #175: Throw an error if cssnext is used with default options not as the first plugin
Some people include cssnext after other plugins, but since cssnext handle import by default, it should be the first so, if possible:

- if cssnext has default options || import to true && import hasn't been piped yet, should throw if not the first plugin
- if cssnext has default options || import to true && import has already been piped, should disable import (& should emit a warning ?) if not the first plugin
- if cssnext has import to false, just check nothing, people probably know what they are doing

That would avoid people doing mistake.

### LINKED ISSUE #180: Clearly show what cssnext includes (autoprefixer, cssnano...)
I am tired of notifying people about duplicate use + missuse of cssnext as a plugin. Ref https://github.com/eggsdesign/egg-incubator/issues/1

And #175 will not help, especially if people add autoprefixer after.

We should add something in the doc + automate some warnings. Ref https://github.com/postcss/postcss/issues/325 poke @ai @ben-eb

### LINKED ISSUE #191: Replace npmignore by exclusion filter in package.json
Ref https://github.com/stylelint/stylelint/pull/333

### LINKED ISSUE #192: Messages can't be disabled in CLI
It would be nice to be able to disable `messages` when using `cssnext` CLI. Also you might want to consider having the `browser` messages disabled by default as this is pretty intrusive and can't be hidden (I'm using a CSS DSL that generates some invalid CSS and always trigger the reporter).

### LINKED ISSUE #199: Add a note for url(...)  with css-modules in webpack.
When you use cssnext with webpack's css-loader in module mode, `url(./file)` is transformed to `url(file)`. This is done by [postcss-url](https://github.com/postcss/postcss-url) included in cssnext. Because of this issue, if you use `url(./image.png)`, the css-loader?module gets it as `url(image.png)` and treats it as a module and searches in moduleDirectories rather than the relative path.

So, with css-loader module mode, I guess it is correct to use `{ url: false }` option in cssnext which is propagated to postcss-url. So can we add a note about this in this repo?

Related issues and discussions:

+ https://github.com/flipkart-incubator/phrontend-webpack/issues/8
+ https://gitter.im/postcss/postcss?at=55d5cd8280eb5af935921b1c
+ https://github.com/boopathi/css-loader-test-2/commit/835eb41255addd7c4506c4e6bfec9d2177cfb08f

### LINKED ISSUE #207: Add "input" and "output" option to config file
It will make npm script written more conveniently without specify input and output files in command line.

### LINKED ISSUE #214: Add postcss-initial
I think we should add this plugin to cssnext https://github.com/maximkoretskiy/postcss-initial

It is `all: initial` polyfill. This property will return all inherited styles to base one.

This plugin is really amazing, because it can be used to create absolutly isolated widgets. Like we will use CSS Modules to isolate selectors and then `all: initial` to isolate pproperties.

### LINKED ISSUE #225: Update pixrem dependency to v3.0.0 for rootValue
Is it possible to update the pixrem dependency to version 3.0.0 so that the rootValue option can be used?

Reference: https://github.com/robwierzbowski/node-pixrem/issues/40

### LINKED ISSUE #230: ⚠︎ Deprecate cssnext in favor or postcss-cssnext
Following #202 and #208

### LINKED ISSUE #231: Outdated Autoprefixer
When should we expect to see Autoprefixer upgraded to 6.0 in cssnext?

### LINKED ISSUE #237: sourcemap can't work
`Your current PostCSS version is 5.0.13, but postcss-color-function uses 4.1.16. Perhaps this is the source of the error below.`
  `events.js:141`
       `throw er; // Unhandled 'error' event`

### PR DESCRIPTION
**Pull request by MoOx**
When ready, this PR will

Closes #173 
Closes #174
Closes #175
Closes #180
Closes #191
Closes #192
Closes #199
Closes #207 
Closes #214 
Closes #225
Closes #231
Closes #237
and finally, closes #230

Since cssnext is now "just" a plugin, this will also closes all the issues related to the tool (cli/api) that are not the responsibility of a plugin:

Closes #73 (see #208 and https://github.com/postcss/postcss/issues/477)
Closes #93
Closes #107
Closes #122
Closes #149 
Closes #170

---

TODO

- [x] gitter update
- [x] doc update for transition
- [x] finish merge/relocate to postcss-cssnext
- [x] update doc with latest features of postcss-cssnext
- [ ] deprecate cssnext and its runners

### CODE REVIEW

**Comment by MoOx** [comment]
Now that everything looks "mergeable", the work to be done is to add a good doc on "how to upgrade from cssnext to postcss-cssnext".

**Comment by MoOx** [comment]
Almost ready, I just need to update the features (some are new) in the doc and then merge/deprecations.

**Comment by magsout** [comment]
:+1: nice work

File: docs/content/postcss.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+These options were just proxy to [PostCSS source map options](https://github.com/postcss/postcss/blob/master/docs/source-maps.md).
+
+#### Examples
+
+If you were using cssnext with default options, you might just need this:
+
+```console
+$ npm uninstall cssnext [--save[-dev]]
+$ npm install postcss postcss-import postcss-url postcss-reporter postcss-browser-reporter postcss-cssnext [--save[-dev]]
```
Comment (magsout) [inline docs/content/postcss.md]: plus `postcss-loader`  no ?

**Comment by magsout** [comment]
Migration documentation is perfect. Good job @MoOx ?

**Pull request merged**